nameko/rpc.py

```diff
@@ -32,69 +36,82 @@
-        if self.queue is None:
 
-            service_name = self.container.service_name
-            queue_name = RPC_QUEUE_TEMPLATE.format(service_name)
-            routing_key = '{}.*'.format(service_name)
+        service_name = self.container.service_name
+        queue_name = RPC_QUEUE_TEMPLATE.format(service_name)
+        routing_key = '{}.*'.format(service_name)
 
-            exchange = get_rpc_exchange(self.container.config)
+        exchange = get_rpc_exchange(self.container.config)
 
-            self.queue = Queue(
-                queue_name,
-                exchange=exchange,
-                routing_key=routing_key,
-                durable=True)
+        self.queue = Queue(
+            queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+            durable=True
+        )
 
-            self.queue_consumer.register_provider(self)
-            self._registered = True
+        config = self.container.config
 
-    def stop(self):
-        """ Stop the RpcConsumer.
+        heartbeat = self.consumer_options.pop(
+            'heartbeat', config.get(HEARTBEAT_CONFIG_KEY, DEFAULT_HEARTBEAT)
+        )
+        prefetch_count = self.consumer_options.pop(
+            'prefetch_count', config.get(
+                PREFETCH_COUNT_CONFIG_KEY, DEFAULT_PREFETCH_COUNT
+            )
+        )
+        accept = self.consumer_options.pop(
+            'accept', serialization.setup(config)[1]
+        )
 
-        The RpcConsumer ordinary unregisters from the QueueConsumer when the
-        last Rpc subclass unregisters from it. If no providers were registered,
-        we should unregister from the QueueConsumer as soon as we're asked
-        to stop.
-        """
-        if not self._providers_registered:
-            self.queue_consumer.unregister_provider(self)
-            self._unregistered_from_queue_consumer.send(True)
+        queues = [self.queue]
+        callbacks = [self.handle_message]
 
-    def unregister_provider(self, provider):
-        """ Unregister a provider.
+        self.consumer = self.consumer_cls(
+            self.amqp_uri, queues=queues, callbacks=callbacks,
+            heartbeat=heartbeat, prefetch_count=prefetch_count,
+            accept=accept
+        )
 
-        Blocks until this RpcConsumer is unregistered from its QueueConsumer,
-        which only happens when all providers have asked to unregister.
-        """
-        self._unregistering_providers.add(provider)
-        remaining_providers = self._providers - self._unregistering_providers
-        if not remaining_providers:
-            _log.debug('unregistering from queueconsumer %s', self)
-            self.queue_consumer.unregister_provider(self)
-            _log.debug('unregistered from queueconsumer %s', self)
-            self._unregistered_from_queue_consumer.send(True)
-
-        _log.debug('waiting for unregister from queue consumer %s', self)
-        self._unregistered_from_queue_consumer.wait()
```

Comment:
This whole "waiting for providers" dance was required to prevent an RPC entrypoint stopping before the RpcConsumer and QueueConsumer that it depended on; if that happened we'd consume a message and then throw `MethodNotFound` exceptions.

We have a test for this (test_rpc_consumer_sharing) and it still passes with the new implementation.